src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/NegotiationProtocol.cs

```diff
@@ -31,7 +32,70 @@ public static void WriteMessage(NegotiationMessage negotiationMessage, Stream ou
             TextMessageFormatter.WriteRecordSeparator(output);
         }
 
-        public static bool TryParseMessage(ReadOnlySpan<byte> input, out NegotiationMessage negotiationMessage)
+        public static void WriteResponseMessage(NegotiationResponseMessage negotiationResponseMessage, Stream output)
+        {
+            using (var writer = new JsonTextWriter(new StreamWriter(output, _utf8NoBom, 1024, leaveOpen: true)))
+            {
+                writer.WriteStartObject();
+                if (!string.IsNullOrEmpty(negotiationResponseMessage.Error))
+                {
+                    writer.WritePropertyName(ErrorPropertyName);
+                    writer.WriteValue(negotiationResponseMessage.Error);
+                }
+                writer.WriteEndObject();
+            }
+
+            TextMessageFormatter.WriteRecordSeparator(output);
+        }
+
+        public static bool TryParseResponseMessage(ReadOnlySpan<byte> input, out NegotiationResponseMessage negotiationMessage)
+        {
+            if (!TextMessageParser.TryParseMessage(ref input, out var payload))
+            {
+                throw new InvalidDataException("Unable to parse payload as a negotiation message.");
+            }
+
+            using (var memoryStream = new MemoryStream(payload.ToArray()))
```

Comment:
There's some code in here in common with `TryParseRequestMessage` as well, can we get that into a private helper?